### PR TITLE
Validate variable types for UniformConstant storage in Vulkan (#2008)

### LINKED ISSUE #2008: Vulkan UniformConstant variables only support opaque types
> Variables identified with the UniformConstant storage class are used only as handles to refer to opaque resources. Such variables must be typed as OpTypeImage, OpTypeSampler, OpTypeSampledImage, or an array of one of these types.

### PR DESCRIPTION
**Pull request by zoddicus**
From the Vulkan 1.1 spec 14.5.2:
  Variables identified with the UniformConstant storage class are used
  only as handles to refer to opaque resources. Such variables must be
  typed as OpTypeImage, OpTypeSampler, OpTypeSampledImage, or an array
  of one of these types.

Fixes #2008

### CODE REVIEW

File: source/val/validate_memory.cpp [hunk 1]
```diff
@@ -323,6 +324,40 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
              << "Classes: Workgroup, CrossWorkgroup, Private, and Function";
     }
   }
+
+  // Vulkan spec 14.5.2
+  if (spvIsVulkanEnv(_.context()->target_env) &&
+      storage_class == SpvStorageClassUniformConstant) {
+    const auto pointer_type_index = 2;
+    const auto pointer_type_id =
+        result_type->GetOperandAs<uint32_t>(pointer_type_index);
```
Comment (dj2) [inline source/val/validate_memory.cpp]: I'd just inline the 2 here (I know we do this in other places, but I don't think it adds much)

File: source/val/validate_memory.cpp [hunk 2]
```diff
@@ -323,6 +324,40 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
...
+    auto variable_type = _.FindDef(pointer_type_id);
+    auto variable_type_opcode = variable_type->opcode();
+
+    // If the variable is actually an array extract the element type.
+    if (variable_type_opcode == SpvOpTypeArray ||
+        variable_type_opcode == SpvOpTypeRuntimeArray) {
+      const auto element_index = 1;
```
Comment (dj2) [inline source/val/validate_memory.cpp]: ditto

File: test/val/CMakeLists.txt [hunk 3]
```diff
@@ -55,6 +55,7 @@ add_spvtools_unittest(TARGET val_ijklmnop
        val_layout_test.cpp
        val_literals_test.cpp
        val_logicals_test.cpp
+       val_memory_test.cpp
```
Comment (dj2) [inline test/val/CMakeLists.txt]: Is this file missing?

File: source/val/validate_memory.cpp (hunk 1 above)
Comment (zoddicus) [reply source/val/validate_memory.cpp]: Done

File: test/val/CMakeLists.txt (hunk 3 above)
Comment (zoddicus) [reply test/val/CMakeLists.txt]: yup, added

**Comment by dj2** [comment]
This LG to me, but leaving the approval to @alan-baker.

File: source/val/validate_memory.cpp [hunk 4]
```diff
@@ -323,6 +324,34 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
...
+
+  // Vulkan spec 14.5.2
+  if (spvIsVulkanEnv(_.context()->target_env) &&
+      storage_class == SpvStorageClassUniformConstant) {
+    auto variable_type = _.FindDef(result_type->GetOperandAs<uint32_t>(2));
+    auto variable_type_opcode = variable_type->opcode();
+
+    // If the variable is actually an array extract the element type.
+    if (variable_type_opcode == SpvOpTypeArray ||
+        variable_type_opcode == SpvOpTypeRuntimeArray) {
```
Comment (alan-baker) [inline source/val/validate_memory.cpp]: According the SPIR-V environment for Vulkan, only SSBOs (StorageBuffer with Block or Uniform with BufferBlock) can have a runtime array so drop this condition.

File: source/val/validate_memory.cpp [hunk 5]
```diff
@@ -323,6 +324,34 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
...
+    }
+
+    switch (variable_type_opcode) {
+      case SpvOpTypeImage:
+      case SpvOpTypeSampler:
+      case SpvOpTypeSampledImage:
+        break;
+      default:
+        return _.diag(SPV_ERROR_INVALID_ID, inst)
+               << "Variables identified with the UniformConstant storage class "
```
Comment (alan-baker) [inline source/val/validate_memory.cpp]: I generally like to mention this is a Vulkan specific check so the user knows where to look.

Comment (zoddicus) [reply source/val/validate_memory.cpp]: Updated the comment at the beginning of this block to be more explicit.

File: source/val/validate_memory.cpp (hunk 4 above)
Comment (zoddicus) [reply source/val/validate_memory.cpp]: Done

File: source/val/validate_memory.cpp (hunk 5 above)
Comment (alan-baker) [reply source/val/validate_memory.cpp]: I meant this error message. The user should have to read this code to know its a Vulkan-specific check.

Comment (zoddicus) [reply source/val/validate_memory.cpp]: my bad, I have added a reference to the spec in the error message

File: test/val/val_memory_test.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+%func = OpFunction %void None %functy
+%1 = OpLabel
+OpReturn
+OpFunctionEnd
+)";
+  CompileSuccessfully(spirv.c_str(), SPV_ENV_VULKAN_1_1);
+  EXPECT_EQ(SPV_ERROR_INVALID_ID, ValidateInstructions(SPV_ENV_VULKAN_1_1));
+  EXPECT_THAT(
+      getDiagnosticString(),
+      HasSubstr("Variables identified with the UniformConstant storage class "
```
Comment (alan-baker) [inline test/val/val_memory_test.cpp]: Please update this to reflect the full message.

File: test/val/val_memory_test.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (alan-baker) [inline test/val/val_memory_test.cpp]: Please update this to reflect the full message.

Comment (zoddicus) [reply test/val/val_memory_test.cpp]: Done

File: test/val/val_memory_test.cpp (hunk 6 above)
Comment (zoddicus) [reply test/val/val_memory_test.cpp]: Done

**Pull request merged**